src/ray/raylet/worker_pool.cc

```diff
@@ -1162,6 +1162,55 @@ void WorkerPool::KillIdleWorker(std::shared_ptr<WorkerInterface> idle_worker,
       });
 }
 
+WorkerUnfitForTaskReason WorkerPool::WorkerFitsForTask(
+    const WorkerInterface &worker, const TaskSpecification &task_spec) const {
+  if (worker.IsDead()) {
+    return WorkerUnfitForTaskReason::OTHERS;
+  }
+  if (worker.GetLanguage() != task_spec.GetLanguage()) {
+    return WorkerUnfitForTaskReason::OTHERS;
+  }
+  // Don't allow worker reuse across jobs or root detached actors. Reuse worker with
+  // unassigned job_id and root detached actor id is OK.
+  JobID job_id = worker.GetAssignedJobId();
+  if (!job_id.IsNil() && job_id != task_spec.JobId()) {
+    return WorkerUnfitForTaskReason::ROOT_MISMATCH;
+  }
+  ActorID root_detached_actor_id = worker.GetRootDetachedActorId();
+  if (!root_detached_actor_id.IsNil() &&
+      root_detached_actor_id != task_spec.RootDetachedActorId()) {
+    return WorkerUnfitForTaskReason::ROOT_MISMATCH;
+  }
+  auto is_gpu = worker.GetIsGpu();
+  if (is_gpu.has_value()) {
+    bool task_is_gpu =
+        task_spec.GetRequiredResources().Get(scheduling::ResourceID::GPU()) > 0;
+    if (is_gpu.value() != task_is_gpu) {
+      return WorkerUnfitForTaskReason::RUNTIME_ENV_MISMATCH;
+    }
+  }
+  auto is_actor_worker = worker.GetIsActorWorker();
+  if (is_actor_worker.has_value() &&
+      is_actor_worker.value() != task_spec.IsActorCreationTask()) {
+    return WorkerUnfitForTaskReason::RUNTIME_ENV_MISMATCH;
+  }
+  // TODO(clarng): consider re-using worker that has runtime envionrment
+  // if the task doesn't require one.
+  if (worker.GetRuntimeEnvHash() != task_spec.GetRuntimeEnvHash()) {
+    return WorkerUnfitForTaskReason::RUNTIME_ENV_MISMATCH;
+  }
+  // Skip if the dynamic_options doesn't match.
+  if (LookupWorkerDynamicOptions(worker.GetStartupToken()) !=
+      task_spec.DynamicWorkerOptionsOrEmpty()) {
+    return WorkerUnfitForTaskReason::DYNAMIC_OPTIONS_MISMATCH;
+  }
+  // These workers are exiting. So skip them.
+  if (pending_exit_idle_workers_.contains(worker.WorkerId())) {
+    return WorkerUnfitForTaskReason::OTHERS;
+  }
+  return WorkerUnfitForTaskReason::NONE;
+}
+
```

Comment:
Could you remove `// TODO(jjyao) This should be renamed to worker cache key hash` from the code as well